### ISSUE TITLE
[AppleBug] UICollectionView reload displaced section crashes in batch update

### ISSUE THREAD

**Issue by rnystrom** [Instagram/IGListKit#690]
Summary:
When reloading a section that is displaced/shifted inside `-[UICollectionView performBatchUpdates:completion:]`, it is possible to hit an `NSInternalInconsistencyException` stating "attempt to perform an insert and a move to the same section (i)" when the sections used in `insertSections:`, `deleteSections:`, `reloadSections:` , and `moveSection:toSection:` do not intersect.

For instance, a collection view who's data source is updated to reflect the following section mutations will crash:

```
[self.collectionView performBatchUpdates:^{
  [self.collectionView moveSection:2 toSection:4];
  [self.collectionView reloadSections:[NSIndexSet indexSetWithIndex:3]];
  [self.collectionView deleteSections:[NSIndexSet indexSetWithIndex:4]];
  [self.collectionView insertSections:[NSIndexSet indexSetWithIndex:2]];
} completion:nil];
```

Steps to Reproduce:
1. Open attached UICollectionViewReloadDisplacement sample app
2. Hit "Play" bar button
3. Crash

Expected Results:
- Section 2 moves to section 4
- Section 3 is reloaded
- Section 4 is deleted
- A new section is inserted into section 2

Actual Results:
Crash with exception:

```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to perform an insert and a move to the same section (4)'
```

Version:
iOS 9.3.4, 8.4, and 10.0 beta 4

Configuration:
iPhone 6, iPhone 6s simulator

**Comment by rnystrom** [comment]
Please dupe

https://openradar.appspot.com/radar?id=5037608161771520

**Comment by PeterAntonsen** [comment]
I think there's something about, deletes and reloads using the index'es of the old-dataset, whereas inserts, uses the index'es of the new-dataset.. so you might want to check up on your relevant indexes as I think this is your issue..

**Comment by rnystrom** [comment]
@PeterAntonsen this is actually a confirmed issue

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by PeterAntonsen** [comment]
@rnystrom, fair enough, just stumbled across this, weighing in from memory as it reminded of another thread I saw.. hope it gets solved :)
